src/file_pipe_log/pipe_builder.rs

```diff
@@ -181,53 +193,125 @@ impl<F: FileSystem> DualPipesBuilder<F> {
+            file_system: self.file_system.clone(),
+            ..append_recovery_cfg
+        };
+        let append_files = &mut self.append_files;
+        let rewrite_files = &mut self.rewrite_files;
+
+        // As the `recover_queue` would update the `Version` of each log file in
+        // `apend_files` and `rewrite_files`, we re-design the implementation on
+        // `recover_queue` to make it compatiable to concurrent processing
+        // with ThreadPool.
         let (append, rewrite) = pool.join(
-            || self.recover_queue(LogQueue::Append, machine_factory, append_concurrency),
-            || self.recover_queue(LogQueue::Rewrite, machine_factory, rewrite_concurrency),
+            || {
+                DualPipesBuilder::recover_queue(
+                    append_recovery_cfg,
+                    append_files,
+                    machine_factory,
+                    append_concurrency,
+                )
+            },
+            || {
+                DualPipesBuilder::recover_queue(
+                    rewrite_recovery_cfg,
+                    rewrite_files,
+                    machine_factory,
+                    rewrite_concurrency,
+                )
+            },
         );
-
         Ok((append?, rewrite?))
     }
 
-    /// Reads through log items in all available log files of the specified
-    /// queue, and replays them to specific [`ReplayMachine`]s that can be
-    /// constructed via `machine_factory`.
+    /// Builds a new storage for the specified log queue.
+    fn build_pipe(&self, queue: LogQueue) -> Result<SinglePipe<F>> {
+        let files = match queue {
+            LogQueue::Append => &self.append_files,
+            LogQueue::Rewrite => &self.rewrite_files,
+        };
+        let first_seq = files.first().map(|f| f.seq).unwrap_or(0);
+        let files: VecDeque<FileHandler<F>> = files
+            .iter()
+            .map(|f| FileHandler {
+                handle: f.handle.clone(),
+                version: f.version,
+                seq: f.seq,
+            })
+            .collect();
+        SinglePipe::open(
+            &self.cfg,
+            self.file_system.clone(),
+            self.listeners.clone(),
+            queue,
+            first_seq,
+            files,
+        )
+    }
+
+    /// Builds a [`DualPipes`] that contains all available log files.
+    pub fn finish(self) -> Result<DualPipes<F>> {
+        let appender = self.build_pipe(LogQueue::Append)?;
+        let rewriter = self.build_pipe(LogQueue::Rewrite)?;
+        DualPipes::open(self.dir_lock.unwrap(), appender, rewriter)
+    }
+
+    /// Returns a mutable reference to the file_list related to `[LogQueue]`.
+    pub fn get_mut_file_list(&mut self, queue: LogQueue) -> &mut Vec<FileHandler<F>> {
+        if queue == LogQueue::Append {
+            &mut self.append_files
+        } else {
+            &mut self.rewrite_files
+        }
+    }
+}
+
+impl<F: FileSystem> DualPipesBuilder<F> {
```

Comment:
Is there a particular reason to separate this into a new block? It can be confusing at first glance.

Reply:
It's just my code-sytle.
I just wanna divide `methods`(called by a specific `struct` object) and `associated functions`(called by the `struct` type) into seperated parts.

Reply:
Clean-up the code-style.